File: packages/core/src/tools/read.ts
import { access as fsAccess, readFile as fsReadFile } from "node:fs/promises";
import { constants } from "node:fs";
import { tool } from "ai";
import { z } from "zod";
import { resolveReadPath } from "./utils/path-utils";
import { recordRead } from "./utils/read-registry";
import { fetchUrlAsText, isHttpUrl } from "./utils/fetch-url";
import { getDoc, renderDocsIndex } from "../docs";
import { DEFAULT_MAX_BYTES, DEFAULT_MAX_LINES, formatSize, truncateHead } from "./utils/truncate";

const IMAGE_EXT_RE = /\.(jpe?g|png|gif|webp)$/i;

function detectImageMimeType(path: string): string | null {
    const m = path.match(IMAGE_EXT_RE);
    if (!m) return null;
    const ext = m[1].toLowerCase();
    if (ext === "jpg" || ext === "jpeg") return "image/jpeg";
    if (ext === "png") return "image/png";
    if (ext === "gif") return "image/gif";
    if (ext === "webp") return "image/webp";
    return null;
}

export interface ReadToolContext {
    cwd: string;
    abortSignal?: AbortSignal;
}

/** Resolve a `loop://` URI: docs index, a specific doc, or a wrapped URL fetch. */
async function readLoopUri(uri: string, signal?: AbortSignal): Promise<string> {
    const rest = uri.slice("loop://".length);
    if (rest === "docs" || rest === "docs/") {
        return renderDocsIndex();
    }
    if (rest.startsWith("docs/")) {
        const name = rest.slice("docs/".length);
        const doc = getDoc(name);
        if (!doc) return `[no such doc: ${name}]\n\n${renderDocsIndex()}`;
        return doc.content;
    }
    if (rest.startsWith("fetch/")) {
        const url = rest.slice("fetch/".length).trim();
        if (!isHttpUrl(url)) return `[loop://fetch/ expects an http(s):// URL, got: ${url}]`;
        return fetchUrlAsText(url, signal);
    }
    return `[unknown loop:// path: ${uri}. Use loop://docs, loop://docs/<name>.md, or loop://fetch/<url>]`;
}

export function createReadTool(ctx: ReadToolContext) {
    return tool({
        description: `Read the contents of a file, fetch a URL, or read loop's internal docs.

- Local path → reads the file (text, truncated to ${DEFAULT_MAX_LINES} lines or ${DEFAULT_MAX_BYTES / 1024}KB; use offset/limit for large files; continue with offset until complete).
- \`loop://fetch/<url>\` (or a bare http(s):// URL) → fetches the page and returns it as readable text (HTML stripped).
- \`loop://docs\` → lists loop's internal docs. \`loop://docs/<name>.md\` reads one (e.g. \`loop://docs/config.md\`).

IMPORTANT: when the user asks to add or change a model, custom provider, hook, MCP server, or custom agent, FIRST read \`loop://docs/config.md\` for the exact file locations and JSON shapes, then make the change. After editing any loop config, tell the user to hard-reload with /reload or by restarting loop — config changes don't apply to the running session until then.`,
        inputSchema: z.object({
            path: z
                .string()
                .describe(
                    "Local file path (relative or absolute), an http(s):// URL, or a loop:// URI (loop://docs, loop://docs/<name>.md, loop://fetch/<url>)",
                ),
            offset: z.number().int().positive().optional().describe("Line number to start reading from (1-indexed)"),
            limit: z.number().int().positive().optional().describe("Maximum number of lines to read"),
        }),
        execute: async ({ path, offset, limit }, options) => {
            const signal = options?.abortSignal ?? ctx.abortSignal;
            if (signal?.aborted) throw new Error("Operation aborted");
            const trimmed = path.trim();
            // loop:// scheme → internal docs or wrapped URL fetch.
            if (trimmed.toLowerCase().startsWith("loop://")) {
                return readLoopUri(trimmed, signal);
            }
            // URL → fetch as text (offset/limit don't apply to web content).
            if (isHttpUrl(path)) {
                return fetchUrlAsText(trimmed, signal);
            }
            const absolutePath = resolveReadPath(path, ctx.cwd);
            await fsAccess(absolutePath, constants.R_OK);
            const mime = detectImageMimeType(absolutePath);
            if (mime) {
                return `[image file: ${mime} at ${absolutePath}. Image rendering not supported in this tool result; use bash 'file' or vision-capable read elsewhere.]`;
            }
            const buf = await fsReadFile(absolutePath);
            // Unlocks edit/write for this file (read-before-modify enforcement).
            recordRead(absolutePath);
            const textContent = buf.toString("utf-8");
            const allLines = textContent.split("\n");
            const totalFileLines = allLines.length;
            const startLine = offset ? Math.max(0, offset - 1) : 0;
            const startLineDisplay = startLine + 1;
            if (startLine >= allLines.length) {
                throw new Error(`Offset ${offset} is beyond end of file (${allLines.length} lines total)`);
            }
            let selectedContent: string;
            let userLimitedLines: number | undefined;
            if (limit !== undefined) {
                const endLine = Math.min(startLine + limit, allLines.length);
                selectedContent = allLines.slice(startLine, endLine).join("\n");
                userLimitedLines = endLine - startLine;
            } else {
                selectedContent = allLines.slice(startLine).join("\n");
            }
            const truncation = truncateHead(selectedContent);
            let outputText: string;
            if (truncation.firstLineExceedsLimit) {
                const firstLineSize = formatSize(Buffer.byteLength(allLines[startLine], "utf-8"));
                outputText = `[Line ${startLineDisplay} is ${firstLineSize}, exceeds ${formatSize(DEFAULT_MAX_BYTES)} limit. Use bash: sed -n '${startLineDisplay}p' ${path} | head -c ${DEFAULT_MAX_BYTES}]`;
            } else if (truncation.truncated) {
                const endLineDisplay = startLineDisplay + truncation.outputLines - 1;
                const nextOffset = endLineDisplay + 1;
                outputText = truncation.content;
                if (truncation.truncatedBy === "lines") {
                    outputText += `\n\n[Showing lines ${startLineDisplay}-${endLineDisplay} of ${totalFileLines}. Use offset=${nextOffset} to continue.]`;
                } else {
                    outputText += `\n\n[Showing lines ${startLineDisplay}-${endLineDisplay} of ${totalFileLines} (${formatSize(DEFAULT_MAX_BYTES)} limit). Use offset=${nextOffset} to continue.]`;
                }
            } else if (userLimitedLines !== undefined && startLine + userLimitedLines < allLines.length) {
                const remaining = allLines.length - (startLine + userLimitedLines);
                const nextOffset = startLine + userLimitedLines + 1;
                outputText = `${truncation.content}\n\n[${remaining} more lines in file. Use offset=${nextOffset} to continue.]`;
            } else {
                outputText = truncation.content;
            }
            return outputText;
        },
    });
}

File: internal/modules/core/tools/read.go
package tools

import (
	"bufio"
	"context"
	"os"
	"strings"
	"unicode/utf8"

	"github.com/notshekhar/pi/internal/modules/ai"
)

const (
	defaultMaxLines = 2000
	defaultMaxBytes = 64 << 10
)

type readArgs struct {
	Path   string `json:"path" jsonschema:"description=Local file path, relative or absolute"`
	Offset *int   `json:"offset,omitempty" jsonschema:"description=1-indexed line to start at,minimum=1"`
	Limit  *int   `json:"limit,omitempty" jsonschema:"description=Maximum number of lines to read,minimum=1"`
}

// Read returns the read tool.
func Read(t *Context) ai.Tool {
	return ai.NewTool("read",
		"Read a file. Large files are truncated; continue with offset= until complete. Directories need ls or glob. Read a file before editing it.",
		func(ctx context.Context, a readArgs) (ai.ToolResult, error) {
			if aborted(ctx) {
				return ai.ToolError("aborted"), nil
			}
			path := Resolve(a.Path, t.CWD)
			info, err := os.Stat(path)
			if err != nil {
				return ai.ToolErrorf("could not read %s: %v", a.Path, err), nil
			}
			if info.IsDir() {
				return ai.ToolErrorf("%s is a directory. Use ls to list it, or glob to match files inside it.", a.Path), nil
			}
			if !utf8SafeFile(path) {
				return ai.ToolErrorf("binary file (%s) at %s. Inspect it with bash (file, xxd, strings).", formatSize(info.Size()), path), nil
			}

			mtime := info.ModTime().UnixNano()
			offset := 1
			if a.Offset != nil && *a.Offset > 0 {
				offset = *a.Offset
			}
			limit := defaultMaxLines
			if a.Limit != nil && *a.Limit > 0 {
				limit = *a.Limit
			}

			f, err := os.Open(path)
			if err != nil {
				return ai.ToolErrorf("could not read %s: %v", a.Path, err), nil
			}
			defer f.Close()

			scan := bufio.NewScanner(f)
			scan.Buffer(nil, 1<<20)
			var (
				lines   []string
				lineNo  int
				bytes   int
				hasMore bool
				sawEOF  = true
				start   int
			)
			for scan.Scan() {
				lineNo++
				if lineNo < offset {
					continue
				}
				if start == 0 {
					start = lineNo
				}
				if len(lines) >= limit || bytes+len(scan.Bytes()) > defaultMaxBytes {
					hasMore = true
					sawEOF = false
					break
				}
				lines = append(lines, scan.Text())
				bytes += len(scan.Bytes()) + 1
			}
			if err := scan.Err(); err != nil {
				return ai.ToolErrorf("reading %s: %v", a.Path, err), nil
			}
			if start == 0 {
				start = offset
			}
			end := start
			if len(lines) > 0 {
				end = start + len(lines) - 1
			}
			t.Registry.RecordRead(path, mtime, start, end, sawEOF && !hasMore)

			text := strings.Join(lines, "\n")
			if hasMore {
				next := end + 1
				text += "\n\n[Showing lines " + itoa(start) + "-" + itoa(end) + ". Use offset=" + itoa(next) + " to continue.]"
			}
			return ai.ToolText(text), nil
		})
}

func utf8SafeFile(path string) bool {
	f, err := os.Open(path)
	if err != nil {
		return false
	}
	defer f.Close()
	buf := make([]byte, 8000)
	n, _ := f.Read(buf)
	if n == 0 {
		return true
	}
	if strings.IndexByte(string(buf[:n]), 0) >= 0 {
		return false
	}
	return utf8.Valid(buf[:n]) || utf8.RuneCount(buf[:n]) > 0
}

func formatSize(n int64) string {
	switch {
	case n < 1024:
		return itoa(int(n)) + "B"
	case n < 1024*1024:
		return itoa(int(n/1024)) + "KB"
	default:
		return itoa(int(n/(1024*1024))) + "MB"
	}
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	var b [20]byte
	i := len(b)
	neg := n < 0
	if neg {
		n = -n
	}
	for n > 0 {
		i--
		b[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		i--
		b[i] = '-'
	}
	return string(b[i:])
}
